lib/ansible/modules/crypto/openssl_pkcs12.py

```diff
@@ -237,7 +237,54 @@ def _check_pkey_passphrase():
         if not state_and_perms:
             return state_and_perms
 
-        return _check_pkey_passphrase
+        if os.path.exists(self.path) and module.params['action'] == 'export':
+            dummy = self.generate(module)
+            self.src = self.path
+            try:
+                pkcs12_privatekey, pkcs12_certificate, pkcs12_other_certificates, pkcs12_friendly_name = self.parse()
+            except crypto.Error:
+                return False
+            if (pkcs12_privatekey is not None) and (self.privatekey_path is not None):
+                expected_pkey = crypto.dump_privatekey(crypto.FILETYPE_PEM,
+                                                       self.pkcs12.get_privatekey())
+                if pkcs12_privatekey != expected_pkey:
+                    return False
+            elif bool(pkcs12_privatekey) != bool(self.privatekey_path):
+                return False
+
+            if (pkcs12_certificate is not None) and (self.certificate_path is not None):
+
+                expected_cert = crypto.dump_certificate(crypto.FILETYPE_PEM,
+                                                        self.pkcs12.get_certificate())
+                if pkcs12_certificate != expected_cert:
+                    for pkcs_cert in pkcs12_other_certificates:
+                        if pkcs_cert == expected_cert:
```

Comment:
Hmm, if the certificate is found in `pkcs12_other_certificates`, I think you still have to return `False`, since you want the certificate to be the leaf certificate and not part of the chain.

Reply:
In my tests, the certs are not necessarily returned in the correct order, sometimes the cert ends up in the `pkcs12_other_certificates` reason why this check. To be fair it does make sense in a way, in the end a pkcs12 file will just be a list of certs bundled together, it's pyopenssl that calls them 'ca_certificates' but they don't need to necessarily be CA certs, they can be any cert.

Reply:
The distinction between cert and CA cert is made by OpenSSL itself: https://www.openssl.org/docs/manmaster/man3/PKCS12_create.html https://www.openssl.org/docs/manmaster/man3/PKCS12_parse.html

That's really strange, since OpenSSL both writes and reads the PKCS12 file, I would expect the order to be preserved.

Reply:
I've been playing around with this a bit; the problem seems to be loading the PKCS12 file with OpenSSL. When simply dumping the result of both runs via `openssl pkcs12 -info`, the output is the same for both. Also, the certificates show up in the correct order, and friendly names are correct. I'll go look at how OpenSSL loads PKCS12 files next...

Reply:
Ok, I think I got it :-) The problem is that in your test, you use the same private key for **all** certificates. This won't happen in real life if you have a bunch of different certificates. The OpenSSL code is here: https://github.com/openssl/openssl/blob/master/crypto/pkcs12/p12_kiss.c

It works as follows:
 - crawl through PKCS12 file and collect private key and all certs (in correct order)
 - pop certs from cert stack (i.e. reverse order!), and:
   - use first cert that matches private key as main cert;
   - if there is no private key, or cert doesn't match private key, or already a cert matching the private key was found, add the cert to CA certs stack.

This results in the following:
 - all the problems you described are caused by all three certs in the .p12 file using the same private key;
 - in general, if no private key is in the .p12 file, friendly name isn't parsed correctly (because the main certificate itself won't be found);
 - the CA certs are returned in reverse order as specified on creation.

This also explains why there was this funny condition that friendly name and private key path were required together (which we removed in #54370): that's because OpenSSL totally screws up otherwise! (I guess this is why that restriction was introduced in the first place. That should have *really* been documented...)

We can obviously work around this, at least partially, but *only* assuming that the handling of pyOpenSSL and OpenSSL of PKCS12 files will never change. This is really !#(*$#@!(*!!@. I guess the main result is "don't use OpenSSL for PKCS12 file handling if possible". Anyway, assuming that pyOpenSSL won't change is pretty safe, since it's mostly deprecated anyway (except for handling TLS connections), this part of the code probably won't change. Well, probably. Who knows. Then there is pyca/pyopenssl#770, in which it is said `"We would accept a PR here to expand support [...]"`.

Reply:
About the naming `ca_certificates` vs. `other_certificates`: how about renaming the option to `other_certificates` (and keeping `ca_certificates` as an alias)? That should reduce confusion a lot.

Reply:
This should hopefully be fixed now